### PR TITLE
[RLlib] Add support for complex observations in CQL

### LINKED ISSUE #23331: [Bug] [RLlib] CQL does not support complex observations
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### Issue Severity

Medium: It contributes to significant difficulty to complete my task but I work arounds and get it resolved.

### What happened + What you expected to happen

When training CQL with complex observations, i.e. structures consisting of `gym.spaces.Dict` or `gym.spaces.Tuple`, `CQLTrainer` throws an error during loss computation

### Versions / Dependencies

Ray Nightly

### Reproduction script

```python
from ray.rllib.agents import cql


# Custom CQLTainer config
config = {}
config['observation_space'] = gym.spaces.Dict({
    'SV': gym.spaces.Box(shape=(1,)),
    'CV': gym.spaces.Box(shape=(3, 200, 200))
})
config['action_space'] = gym.spaces.Box(shape=(2,))
config['Q_model'] = {
    'custom_model': 'ComplexInputNetwork',  # Taken from docs at https://docs.ray.io/en/latest/rllib/rllib-models.html#more-examples-for-building-custom-models
    'conv_filters': [[4, 5, 1], [8, 10, 1]],
    'fcnet_hiddens': [256, 256, 256],
}
config['policy_model'] = {
    'custom_model': 'ComplexInputNetwork',
    'conv_filters': [[4, 5, 1], [8, 10, 1]],
    'fcnet_hiddens': [256, 256, 256],
}
config['input'] = '/path/to/my_offline_rl_data.json'


trainer = cql.CQLTrainer(config=config)
for i in range(10):
    trainer.train()
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by fwitter**
## Why are these changes needed?

When training CQL with complex observations, i.e. structures consisting of `gym.spaces.Dict` or `gym.spaces.Tuple`, `CQLTrainer` throws an error during loss computation. This PR resolves the problem.

## Related issue number

Closes #23331

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/cql/cql_torch_policy.py [hunk 1]
```diff
@@ -107,6 +108,8 @@ def cql_loss(
     next_obs = train_batch[SampleBatch.NEXT_OBS]
     terminals = train_batch[SampleBatch.DONES]
 
+    batch_size = tree.flatten(obs)[0].shape[0]
```
Comment (gjoliver) [inline rllib/agents/cql/cql_torch_policy.py:111]: can you move this lower to be right above the first if statement that needs it?

File: rllib/agents/cql/cql_tf_policy.py [hunk 2]
```diff
@@ -46,30 +47,34 @@
 MEAN_MAX = 9.0
 
 
+def _repeat_tensor(t, n):
+    t_rep = tf.expand_dims(t, 1)
```
Comment (gjoliver) [inline rllib/agents/cql/cql_tf_policy.py:51]: can you add comments to explain the tensor maneuver here? otherwise the logics may be hard to follow.

**Review by gjoliver — COMMENTED** [review]
any chance you can help update the torch policy as well?
thanks.

File: rllib/agents/cql/cql_tf_policy.py (hunk 2 above)
Comment (fwitter) [reply rllib/agents/cql/cql_tf_policy.py:52]: I added some comments but actually I just extended the existing code to work with an arbitrary number of dimensions. Not sure why the repetition process is like it is.

**Comment by fwitter** [comment]
> any chance you can help update the torch policy as well? thanks.

Already included :)

**Review by gjoliver — APPROVED** [review]
Thanks for the PR.
@sven1977, can you help merge?

**Comment by avnishn** [comment]
how have you been checking to see that this works @fwitter ?

could we add a small test to `test_cql.py`?
